logger/unilog.go

```diff
@@ -149,6 +153,83 @@ const (
 // Stats is Unilog's statsd client.
 var Stats *statsd.Client
 
+// tagPair is a simple pair of a tag t and the full metric name n
+type tagPair struct {
+	t string
+	n string
+}
+
+// independentTags stores a list of tags to individually emit metrics on.
+// Each metric name will be formatted exactly once and cached in metricsTable.
+type independentTags struct {
+	// The tags to build metric names from
+	// Format is foo:bar where foo is the tag name
+	Tags []string
+	// Lookup table for metricName -> slice of metricName.tagName
+	metricsTable map[string][]tagPair
+}
+
+func newIndependentTags(tags []string) *independentTags {
+	return &independentTags{Tags: tags, metricsTable: make(map[string][]tagPair)}
+}
+
+func setupIndependentTags() *independentTags {
+	return newIndependentTags(strings.Split(independenttags, ","))
+}
+
+func (it *independentTags) GetTags(key string) []tagPair {
+	if it == nil {
+		return []tagPair{}
+	}
+	tags, ok := it.metricsTable[key]
+	if ok {
+		return tags
+	}
+	for _, tag := range it.Tags {
+		prefix := strings.Split(tag, ":")[0]
+		// If we don't get a token (e.g. we are passed the empty string), skip this tag
+		if len(prefix) == 0 {
+			continue
+		}
+		tags = append(tags, tagPair{t: tag, n: fmt.Sprintf("%s.%s", key, prefix)})
+	}
+	it.metricsTable[key] = tags
+	return tags
+}
+
+// tagState holds the state necessary to efficiently emit independent metrics
+var tagState *independentTags
+
+// Client is the interface for our metrics client for use in independent metric emission
+// Currently, only IndependentCount is implemented
+type Client interface {
+	Count(name string, value int64, tags []string, rate float64) error
+}
+
+// IndependentCount is a wrapper for the statsd.Count method. It will emit the normal metric
+// in addition to a metric for each tag in independenttags with all global tags and that tag
+// attached (along with tags passed as an argument to IndependentCount).
+// Metric names will be of the form metricName.tag. Will short-circuit upon encountering an error.
+func IndependentCount(client Client, name string, value int64, tags []string, rate float64) error {
+	// Preserve backwards compatability by emitting the normal metric
+	err := client.Count(name, value, tags, rate)
+	if err != nil {
+		return err
+	}
+	names := tagState.GetTags(name)
```

Comment:
This might be clearer as something like `pairs := tagState.getTags(name)`; `names` is a bit confusing

Reply:
👍